Add null check to getCPtr

ajsutton opened the pull request:
When verifying aggregate signatures using the Java wrapper there's a sequence of calls to `Pairing.aggregate`, the first of which has all non-null values but subsequent calls should pass `null` for the `P2_Affline sig` parameter. In Teku this comes out as https://github.com/ConsenSys/teku/blob/d585e7f1dc25333defe92e9f1d47b08b122533a5/bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java#L78 that method being called in a loop by the `verify(List<PublicKeyMessagePair>)` method a little below.

However, because `Pairing.aggregate` calls `P2_Affine.getCPtr(sig)` a `NullPointerException` results when `sig` is `null`.  If my understanding is right, `getCPtr` should return a 0 length `long[]` to represent null.  I've made that change (used a static final to avoid allocating a new empty array on each call) and the Teku tests are now passing.

I'm more than happy to be told there's a better way of doing this though.

Nashatyrev commented:
Yep, the previous version generated that code: 
```java
  protected static long getCPtr(P2_Affine obj) {
    return (obj == null) ? 0 : obj.swigCPtr;
  }
```
the current version generates: 
```
  protected static long[] getCPtr(P1_Affine obj) { return obj.swigCPtr; }
```

dot-asm commented:
Wouldn't `{ return (obj == null) ? null : obj.swigCPtr; }` do the job?

dot-asm commented:
It boils down to question how C side would have to handle it. Java `null` does translate to C `NULL`, and C side knows what to do with it. While if you pass `long[0]`, it would be a non-NULL object on C side, and blst would assume that there is specific amount of words... In other words `(obj == null) ? null : obj.swigCPtr` ought to be the right answer...

dot-asm commented:
> While if you pass `long[0]`, it would be a non-NULL object on C side, and [as it stands now] blst would assume that there is specific amount of words...

Just in case for reference. In the context "blst" refers even to JNI wrapper. And if we're to settle for `long[0]`, then we would have to adjust JNI to recognize it and translate it to NULL.

To summarize, check that  `(obj == null) ? null : obj.swigCPtr` works in actual application, then adjust this PR accordingly. Or just say that it works and I'll do the adjustments for you.

dot-asm commented:
> My first thought was to return null but that caused a crash.

Where? In your Java code or in JNI/C? If latter, is there stack back trace?

dot-asm commented:
I can imagine that `long[0]` works, but I'm not convinced that it's not circumstantial. JNI gets an array and if you ask how long is it, it says 0 elements. But JNI doesn't ask, it just goes ahead and accesses as if it's complete point. Now, if all those out-of-bound longs are 0, then it will work. But if they are not, it would be devastating. Is there guarantee that there will be sufficient amount of 0-s?

ajsutton commented:
Sorry, was on my phone so didn't have the full details.  If we change getCPtr to:
```
  protected static long[] getCPtr($javaclassname obj) { return obj != null ? obj.swigCPtr : null; }
```

such that a Java null winds up being passed for `sig`, that Java null should be translated to a `NULL` on the native side.  But then I get:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000108772704, pid=16889, tid=9731
#
# JRE version: OpenJDK Runtime Environment AdoptOpenJDK (14.0.2+12) (build 14.0.2+12)
# Java VM: OpenJDK 64-Bit Server VM AdoptOpenJDK (14.0.2+12, mixed mode, sharing, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# V  [libjvm.dylib+0x772704]  EnableIf<HasDecorator<1097844ull, MO_UNORDERED>::value, oopDesc*>::type RawAccessBarrier<1097844ull>::load_internal<1097844ull, oopDesc*>(void*)+0x4
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/aj/Documents/code/teku/bls/hs_err_pid16889.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/AdoptOpenJDK/openjdk-support/issues
#

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

I'm not at all familiar with the native side of this but based on what you're saying it sounds like it doesn't explicitly check the array length or perform a `NULL` check and one of them seems to be required.  I can get a full core dump for you if it's useful.

dot-asm commented:
> or perform a `NULL` check

NULL is handled at lowest level, i.e. lower that JNI. In other words JNI should not check for NULL, but simply pass it down.

As for complete core. Well, it does mention problematic frame and it's in jvm itself, but it might be appropriate to double-confirm it. So do what it says about ulimit, then run `lldb ``which java`` core`, and finally `bt` at lldb prompt.

Is it possible that your JVM simply refuses to pass null's to JNI? I can confirm that mine can... Can it be version specific?

dot-asm commented:
> I can confirm that mine can...

I've checked with byte[] and it didn't crash, but let me check with long[]. It might happen that NULL has to be handled in a typemap...

ajsutton commented:
Pretty sure all JVM's should be able to pass `null` safely.  I've got a bunch of morning meetings now but I'lll take some time today and put together a simple test case I can send you, plus look into the core dump for you.

dot-asm commented:
It's OK, as I'm going offline soon, so you'll have whole day. Though I might make a suggestion for you to test...

dot-asm commented:
Before you dive into it, try following with `(obj == null) ? null : obj.swigCPtr`
```
--- a/bindings/blst.swg
+++ b/bindings/blst.swg
@@ -244,6 +244,9 @@ import java.nio.file.*;
 %typemap(in)            SWIGTYPE&, SWIGTYPE* %{
     $1 = ($1_ltype)JCALL(GetLongArrayElements, $input, 0);
 %}
+%typemap(in)      const SWIGTYPE&, const SWIGTYPE* %{
+    if ($input) $1 = ($1_ltype)JCALL(GetLongArrayElements, $input, 0);
+%}
 %typemap(out)           SWIGTYPE&, SWIGTYPE* %{
     if ($1 != $null) {
         size_t sz = (sizeof($1_basetype) + sizeof(jlong) - 1)/sizeof(jlong);
@@ -263,7 +266,7 @@ import java.nio.file.*;
     JCALL(ReleaseLongArrayElements, $input, (jlong *)$1, 0);
 %}
 %typemap(freearg) const SWIGTYPE&, const SWIGTYPE* %{
-    JCALL(ReleaseLongArrayElements, $input, (jlong *)$1, JNI_ABORT);
+    if ($input) JCALL(ReleaseLongArrayElements, $input, (jlong *)$1, JNI_ABORT);
 %}
 %typemap(freearg) const std::string& ""
 
```
Out for today. Cheers.

ajsutton commented:
> Before you dive into it, try following with `(obj == null) ? null : obj.swigCPtr`

That patch makes it work, thanks.  I've added it into the PR.  Thanks so much for your rapid assistance with this - as you've probably gathered, my first encounter with SWIG was yesterday so I have no idea what I'm doing. :)

In terms of reproducing the crash if for some reason you need to, I just added an extra line to runnable.java:

```
var ctx = new Pairing(true, DST);
ctx.aggregate(_pk, _sig, msg, pk_for_wire);
ctx.aggregate(_pk, null, msg, pk_for_wire);  // This is new
ctx.commit();
```

Obviously the signature is no longer expected to be valid but we should get a "disaster" exception and not a crash.

dot-asm commented:
I did some reformatting and fixed similar problem with `const byte *`. Thanks for report!

On note related to quoted ctx.aggregate call in the beginning. You should be able to omit `new byte[0]` (or pass `null`), since there is augmentation-free method.

dot-asm commented:
Additional thanks as JavaScript and Python needed corresponding adjustments as well:-)
